ARCs/arc-0020.md

```diff
@@ -0,0 +1,870 @@
+to the controlling Smart Contract. This MAY be performed at two separate times.
+
+The reminder of this section is non-normative.
+
+> Smart ASAs SHOULD NOT require users to opt in to the controlling Smart
+> Contract, unless the implementation requires storing information into their
+> local schema (for instance, to implement [freezing](#asset-freeze); also see
+> [security considerations](#security-considerations)).
+>
+> Clients MAY inspect the local state schema of the controlling Smart Contract
+> to infer whether opt in is required.
+>
+> If a Smart ASA requires opt in, then clients SHOULD prevent users from closing
+> out the controlling Smart Contract unless they don't hold an ASA balance.
+
+## Rationale
+
+This ARC builds on the strengths of the ASA to enable a Smart Contract to
+control its operations and flexibly re-configure its configuration.
+
+The rationale is to have a "Smart ASA" that is as widely adopted as the ASA both
+by the community and by the surrounding ecosystem. Wallets and dApps:
+
+- Will display a user's Smart ASA balance out-of-the-box (because of the
+  underlying ASA).
+- SHOULD recognize Smart ASAs and inform the users accordingly.
+- SHOULD enable users to transfer the Smart ASA by constructing the appropriate
+  transactions, which call the ABI methods of the controlling Smart Contract.
+
+With this in mind, this standard optimizes for:
+
+- Community adoption, by minimizing the [ASA metadata](#metadata) that need to
+  be set and the requirements of a conforming implementation.
+- Developer adoption, by re-using the familiar ASA transaction reference in the
+  methods' specification.
+- Ecosystem integration, by minimizing the amount of work that a wallet, dApp or
+  service should perform to support the Smart ASA.
+
+## Backwards Compatibility
+
+Existing ASAs MAY adopt this standard if issued or re-configured to match the
+requirements in the [metadata section](#metadata).
+
+This requires:
+
+- The ASA to be `DefaultFrozen`.
+- Deploying a Smart Contract that will manage, control and operate on the
+  asset(s).
+- Re-configuring the ASA, by setting its `ClawbackAddr` to the account of the
+  controlling Smart Contract.
+- Associating the ID of the Smart Contract to the ASA (see
+  [metadata](#metadata)).
+
+### ARC-18
+
+Assets implementing [ARC-18](./arc-0018.md) MAY also be compatible with this ARC
+if the Smart Contract implementing royalties enforcement exposes the ABI methods
+specified here. The corresponding ASA and their metadata are compliant with this
+standard.
+
+## Reference Implementation
+
+TODO.
+
+## Security Considerations
+
+Keep in mind that the rules governing a Smart ASA are only in place as long as:
+
+- The ASA remains frozen;
+- the `ClawbackAddr` of the ASA is set to a controlling Smart Contract, as
+  specified in the [metadata section](#metadata);
+- the controlling Smart Contract is not updatable, nor deletable, nor
```

Comment:
Suggestion(?):

> - The controlling Smart Contract is a superset of the ABI defined by this standard

Reply:
I am conflicted here; even if the controlling SC implements the ABI, that doesn't mean that rules won't change (e.g. if you updated it, etc).